lib/CMakeLists.txt

```
@@ -0,0 +1,178 @@
+	when_all_task.hpp
+	get_awaiter.hpp
+	is_awaiter.hpp
+	any.hpp
+	sync_wait_task.hpp
+	unwrap_reference.hpp
+	lightweight_manual_reset_event.hpp
+)
+list(TRANSFORM detailIncludes PREPEND "${PROJECT_SOURCE_DIR}/include/cppcoro/detail/")
+
+set(privateHeaders
+	cancellation_state.hpp
+	socket_helpers.hpp
+	auto_reset_event.hpp
+	spin_wait.hpp
+	spin_mutex.hpp
+)
+
+set(sources
+	async_auto_reset_event.cpp
+	async_manual_reset_event.cpp
+	async_mutex.cpp
+	cancellation_state.cpp
+	cancellation_token.cpp
+	cancellation_source.cpp
+	cancellation_registration.cpp
+	lightweight_manual_reset_event.cpp
+	ip_address.cpp
+	ip_endpoint.cpp
+	ipv4_address.cpp
+	ipv4_endpoint.cpp
+	ipv6_address.cpp
+	ipv6_endpoint.cpp
+	static_thread_pool.cpp
+	auto_reset_event.cpp
+	spin_wait.cpp
+	spin_mutex.cpp
+)
+
+if(WIN32)
+	set(win32DetailIncludes
+		win32.hpp
+		win32_overlapped_operation.hpp
+	)
+    list(TRANSFORM win32DetailIncludes PREPEND "${PROJECT_SOURCE_DIR}/include/cppcoro/detail/")
+    list(APPEND detailIncludes ${win32DetailIncludes})
+
+    set(win32NetIncludes
+        socket.hpp
+        socket_accept_operation.hpp
+        socket_connect_operation.hpp
+        socket_disconnect_operation.hpp
+        socket_recv_operation.hpp
+        socket_recv_from_operation.hpp
+        socket_send_operation.hpp
+        socket_send_to_operation.hpp
+    )
+    list(TRANSFORM win32NetIncludes PREPEND "${PROJECT_SOURCE_DIR}/include/cppcoro/net/")
+    list(APPEND netIncludes ${win32NetIncludes})
+
+    set(win32Sources
+        win32.cpp
+        io_service.cpp
+        file.cpp
+        readable_file.cpp
+        writable_file.cpp
+        read_only_file.cpp
+        write_only_file.cpp
+        read_write_file.cpp
+        file_read_operation.cpp
+        file_write_operation.cpp
+        socket_helpers.cpp
+        socket.cpp
+        socket_accept_operation.cpp
+        socket_connect_operation.cpp
+        socket_disconnect_operation.cpp
+        socket_send_operation.cpp
+        socket_send_to_operation.cpp
+        socket_recv_operation.cpp
+        socket_recv_from_operation.cpp
+    )
+    list(APPEND sources ${win32Sources})
+
+    list(APPEND libraries Ws2_32 Mswsock Synchronization)
+    list(APPEND compile_options /EHsc)
+
+    if("${MSVC_VERSION}" VERSION_GREATER_EQUAL 1900)
+      # TODO remove this when experimental/non-experimental include are fixed
+      list(APPEND compile_definition _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING=1)
+    endif()
+endif()
+
+add_library(cppcoro
+	${includes}
+	${netIncludes}
+	${detailIncludes}
+	${privateHeaders}
+	${sources}
+)
+
+target_include_directories(cppcoro PUBLIC
+	$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+	$<INSTALL_INTERFACE:include>)
+target_compile_features(cppcoro PUBLIC cxx_std_20)
+target_compile_definitions(cppcoro PUBLIC ${compile_definition})
+target_compile_options(cppcoro PUBLIC ${compile_options})
+
+find_package(Coroutines COMPONENTS Experimental Final REQUIRED)
```

This fails with VS 2019 16.8.2. The find_package(Coroutines ...) reports this error: 
> CMake Error at cmake/FindCoroutines.cmake:272 (message):
>   Cannot compile simple program using std::coroutines. Is C++17 or later activated?

Since 3 lines above you specifiy for this target c++20/latest, you should activate it also for the check in the FindCoroutines. This does the trick: 
```suggestion
set(CMAKE_CXX_STANDARD 20)
find_package(Coroutines COMPONENTS Experimental Final REQUIRED)
```
With this change, the CMake generation is OK, but now you will get compilation errors, because now you are using the 'real' c++20 coroutines and not the TS ones. 
> C1189 The <experimental/coroutine> and <experimental/resumable> headers are only supported with /await and implement pre-C++20 coroutine support. Use <coroutine> for standard C++20 coroutines.

The solution is thus to merge also the changes from https://github.com/andreasbuhr/cppcoro/tree/unify_experimental_includes
The master branch from andreasbuhr, that combines this various changes, generates and compiles fine, after the patching with the suggested change `set(CMAKE_CXX_STANDARD 20)` before `find_package(Coroutines ...)`

---

Unfortunately, the fix you propose is not possible. As you pointed out, MSVC 2019 supports both the coroutine-ts and C++20 coroutines. If the user wants to build her project with the coroutine-ts, also cppcoro has to be built with the coroutine-ts. Hardcoding CMAKE_CXX_STANDARD to "20" would make that impossible.
Only the user knows whether she wants the coroutine-ts and C++17 or C++20 coroutines. There is no way of detecting that. So there is no other way than to ask the user what she wants when building. This is what the error message "Cannot compile simple program using std::coroutines. Is C++17 or later activated?" does.
cppcoro needs to be build using either "cmake -DCMAKE_CXX_STANDARD=17" OR "cmake -DCMAKE_CXX_STANDARD=20". This, however, needs to be documented better.

Would you agree with my analysis?

---

I thought that this decision (coroutine-ts and C++17 or C++20 coroutines) is already hardcoded 3 lines above `target_compile_features(cppcoro PUBLIC cxx_std_20)`. I will test your proposal with the invocation of cmake using the CMAKE_CXX_STANDARD define. But I have a bad gut feeling, especially since I hoped that your version would once also be incorporated in vcpkg. The current port on vcpkg has currently the TS version hardcoded. I do not know if with vcpkg you can let the user choose the cxx_standard version.

---

Thanks a lot for poiniting this out:
target_compile_features(cppcoro PUBLIC cxx_std_20)
this has to be fixed.
I hope I can add C++17/coroutine-ts to the Github actions CI in andreasbuhr/cppcoro in the coming weekend, then we catch these things.
Thanks also to pointing to the issue with vcpkg. I will look into that in the coming weekend.

---

> Only the user knows whether she wants the coroutine-ts and C++17 or C++20 coroutines.

The devs who already jumped on the experimental coroutines bandwagoon are leaving on the edge anyway, ahead of HEAD, taking some risks and enjoing the latest and greatest :-) My personal opinion, is that the old stuff in this area can be deprecated, as soon as you have new working compiler version officially available.

---

I removed the statement now from this pull request.

About your suggestion about deprecation: I would agree, but I'd like to have one release of cppcoro which works with the coroutine-ts and C++17 first. After the release, we can remove the C++17 / coroutine-ts support. That way we get it out of the way in the long run and at the same time we can direct people who need coroutine-ts support to that release.